File: app/src/main/java/dev/dimension/flare/ui/screen/status/TwitterArticleScreen.kt
package dev.dimension.flare.ui.screen.status

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ListItemDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.unit.dp
import dev.dimension.flare.model.AccountType
import dev.dimension.flare.ui.component.BackButton
import dev.dimension.flare.ui.component.FlareLargeFlexibleTopAppBar
import dev.dimension.flare.ui.component.FlareScaffold
import dev.dimension.flare.ui.component.RichText
import dev.dimension.flare.ui.component.SubcomposeNetworkImage
import dev.dimension.flare.ui.component.listCard
import dev.dimension.flare.ui.component.placeholder
import dev.dimension.flare.ui.model.ClickContext
import dev.dimension.flare.ui.model.UiState
import dev.dimension.flare.ui.model.onError
import dev.dimension.flare.ui.model.onLoading
import dev.dimension.flare.ui.model.onSuccess
import dev.dimension.flare.ui.presenter.home.xqt.TwitterArticlePresenter
import dev.dimension.flare.ui.presenter.invoke
import dev.dimension.flare.ui.screen.settings.AccountItem
import dev.dimension.flare.ui.theme.screenHorizontalPadding
import dev.dimension.flare.ui.theme.single
import moe.tlaster.precompose.molecule.producePresenter

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun TwitterArticleScreen(
    accountType: AccountType,
    tweetId: String,
    articleId: String?,
    onBack: () -> Unit,
) {
    val uriHandler = LocalUriHandler.current
    val scrollState = rememberScrollState()
    val topAppBarScrollBehavior = TopAppBarDefaults.exitUntilCollapsedScrollBehavior()
    val state by producePresenter(key = "$accountType-$tweetId-$articleId") {
        TwitterArticlePresenter(accountType, tweetId, articleId).invoke()
    }
    FlareScaffold(
        modifier = Modifier.nestedScroll(topAppBarScrollBehavior.nestedScrollConnection),
        topBar = {
            FlareLargeFlexibleTopAppBar(
                title = {
                    state.data
                        .onSuccess {
                            Text(
                                text = it.title,
                            )
                        }.onLoading {
                            Text(
                                text = "Loading...",
                                modifier = Modifier.placeholder(true),
                            )
                        }
                },
                navigationIcon = { BackButton(onBack) },
                scrollBehavior = topAppBarScrollBehavior,
            )
        },
    ) { contentPadding ->
        Column(
            modifier =
                Modifier
                    .verticalScroll(scrollState)
                    .padding(horizontal = screenHorizontalPadding)
                    .padding(contentPadding)
                    .padding(vertical = 16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            state.data
                .onSuccess { data ->
                    data.image?.let { image ->
                        SubcomposeNetworkImage(
                            model = image,
                            contentDescription = data.title,
                            modifier =
                                Modifier
                                    .fillMaxWidth()
                                    .height(220.dp)
                                    .clip(
                                        ListItemDefaults.single().shape,
                                    ),
                        )
                    }
                    AccountItem(
                        userState = UiState.Success(data.profile),
                        onClick = {
                            data.profile.onClicked(
                                ClickContext(
                                    launcher = { uri -> uriHandler.openUri(uri) },
                                ),
                            )
                        },
                        toLogin = {},
                        shapes = ListItemDefaults.single(),
                    )
                    SelectionContainer {
                        RichText(
                            text = data.content,
                        )
                    }
                }.onLoading {
                    Text(
                        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam elementum eget dui a bibendum. Fusce eget porttitor est, et rhoncus massa. Etiam cursus urna at odio vulputate semper. Interdum et malesuada fames ac ante ipsum primis in faucibus. Quisque tincidunt rhoncus massa sed volutpat. Nulla porta orci et finibus accumsan. Duis maximus diam quis congue suscipit. Suspendisse velit enim, mollis non tellus eu, auctor vulputate diam. Sed ut purus eleifend, tempor lectus ac, imperdiet tellus. Proin eleifend lorem ut risus gravida, id bibendum metus posuere. Cras pretium tortor mi. Quisque ac congue urna. Morbi posuere ac orci vestibulum euismod. Maecenas venenatis, justo at aliquet venenatis, arcu mauris sodales ligula, a iaculis nulla eros at turpis. Quisque varius lobortis porttitor.",
                    )
                }.onError {
                    Text(
                        text = it.message ?: "Failed to load article",
                    )
                }
        }
    }
}

File: iosApp/flare/UI/Screen/TwitterArticleScreen.swift
import SwiftUI
import KotlinSharedUI

struct TwitterArticleScreen: View {
    @StateObject private var presenter: KotlinPresenter<TwitterArticlePresenterState>
    
    let placeholderText = """
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam consectetur sapien a tortor sollicitudin fermentum. Vestibulum at tincidunt ipsum. In maximus justo dui, ut auctor ante faucibus vitae. Donec fermentum nec sem id viverra. Donec efficitur feugiat consectetur. Cras tempor suscipit tempus. Sed velit elit, vestibulum ut rhoncus in, pretium eget justo. Integer nec vulputate nisi.

Nullam eu augue ut est gravida pharetra. Cras ullamcorper sodales enim, eget feugiat metus aliquam in. Donec elementum venenatis nisi, non hendrerit sem commodo bibendum. Sed pharetra luctus ipsum, vitae lacinia sem. Proin sodales, enim sed vestibulum semper, nisi ipsum sagittis ante, at semper elit purus in ipsum. Fusce accumsan varius scelerisque. Pellentesque blandit risus purus, id convallis sem fermentum nec. Donec accumsan ullamcorper porta.

Vestibulum tempus turpis nibh. Praesent tempus varius mattis. Nullam et cursus diam, sit amet tincidunt quam. Nunc id nunc nibh. Etiam vel gravida metus. Quisque eget lacus hendrerit, aliquam diam non, faucibus dui. Praesent dignissim tortor in rutrum euismod. Pellentesque elit lorem, imperdiet sed lectus sed, venenatis molestie urna.

Mauris porttitor sapien ex, sed pharetra nibh mattis id. Interdum et malesuada fames ac ante ipsum primis in faucibus. Proin euismod congue risus, ut fermentum mi. Pellentesque sagittis commodo malesuada. Cras molestie vulputate nisl quis placerat. Integer egestas imperdiet sem, ac bibendum tortor imperdiet sed. Pellentesque maximus velit quis felis interdum tincidunt.

Maecenas fringilla vitae leo sit amet lacinia. Donec in dui a ex hendrerit volutpat. Etiam sit amet aliquet arcu. Phasellus placerat at eros eu ornare. Morbi venenatis mi sed tortor aliquet, nec volutpat ex commodo. In interdum elit ac leo efficitur, ultricies sodales enim feugiat. Fusce rutrum erat felis, malesuada varius risus imperdiet vitae. Ut bibendum sagittis metus, a placerat ex varius vel.

"""
    let accountType: AccountType
    let tweetId: String
    let articleId: String?
    
    init(accountType: AccountType, tweetId: String, articleId: String?) {
        self.accountType = accountType
        self.tweetId = tweetId
        self.articleId = articleId
        self._presenter = .init(wrappedValue: .init(presenter: TwitterArticlePresenter(accountType: accountType, tweetId: tweetId, articleId: articleId)))
    }
    
    var body: some View {
        StateView(state: presenter.state.data) { article in
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    if let image = article.image {
                        ListCardView {
                            NetworkImage(data: image)
                                .frame(maxHeight: 280)
                                .clipShape(.rect(cornerRadius: 12))
                        }
                    }
                    Text(article.title)
                        .font(.title)
                        .bold()
                        .frame(maxWidth: .infinity, alignment: .leading)
                    UserCompatView(data: article.profile)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    RichText(text: article.content)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                .padding()
            }
        } errorContent: { error in
            Text(error.message ?? "Failed to load article")
        } loadingContent: {
            ScrollView {
                VStack {
                    Text("Loading...")
                        .font(.title)
                        .bold()
                        .redacted(reason: .placeholder)
                    ListCardView {
                        Text(placeholderText)
                            .redacted(reason: .placeholder)
                            .padding()
                    }
                }
                .padding()
            }
        }
        .navigationBarTitleDisplayMode(.inline)
    }
}
